SNR-based dynamic power system, refactor to new unit

CapnBry opened the pull request:
This replaces the existing RSSI/LQ based dynamic power system with an SNR/LQ based algorithm where possible. All the dynamic power code was also moved out of tx_main to a new unit. Benefits are 5x faster reaction to changes, proactive power bumps, multiple power increases per cycle, and per-rate thresholds.

## Porblem
RSSI turns out to not be the greatest way to implement dynamic power. When the RX has a LNA on board, this boosts the RSSI seen at the LoRa chip and messes up the static up/down thresholds. Through [rigorous testing with multiple different types receivers at the same time](https://heatermeter.com/elrs/), and Sunjun's slicing and dicing of the data, it was determined that SNR can be a better indicator of link strength.

![image](https://user-images.githubusercontent.com/677183/163466228-87bd56ad-43b2-496c-b800-b436d70f1955.png)

In addition, the current RSSI-based system takes 5x telemetry packets to make a single up/down decision. This is needed because the RX only sends the smoothed last RSSI, which is noisy so some averaging is needed to prevent the power from changing on every LinkStats telemetry packet.

Finally, the existing system relies on incoming LinkStats to make any incremental power change, meaning if the signal is blocked, the TX waits until the connection drops entirely before boosting the power (to MAX). This can reestablish the connection, but the RX may have already failsafed by the time the TX reaches this state.

## New Hotness
The new system changes both the RX and TX sides. 

### SNR averaging on the RX
The only change to the RX: average the SNR value between each LinkStats packet (straight average, not moving average) and send that in the telemetry instead of just the last SNR. This creates the smoothing at the RX that was previously attempted on the TX side.

### SNR precision increased
The SNR provided by the Radio is in 0.25dB scale, but we currently discard the fractional part. This changes that to preserve the whole precision and only report the truncated value to OpenTX or the FC. The higher precision means a decent SNR is still available at lower packet rates.

### Decide based on a single telemetry packet
The smoothing being done on a much larger dataset means that the TX side can now make a dynamic power decision **on a single telemetry packet**. This is the biggest feature of this change in my opinion, as the now power adjusts within 512ms (given a minimum telemetry ratio) instead of the old 2.5s.

### Multiple power changes per telemetry packet
The SNR value is also much more reliable, allowing the TX to up the power multiple steps at once from that single packet. The power will only be reduced at most one level per packet though, to be as cautious as possible, and the LQ must be at or above 95% for the power to be lowered.

### Proactive power increase
If armed, the TX now also will increase the power one level if a LinkStats telemetry packet is missed, and then one more level for **any** telemetry packet missed after that until telemetry resumes. For "standard" telemetry rate (4Hz) that means one power bump at 512ms, then one at 768ms, 1024ms etc untill full power is reached. This means that the TX will already be at full power before going to the disconnected state, rather than waiting until disconnect to jump the power up all at once.

### Per-rate up/down thresholds
The TX adds up and down SNR thresholds for each Air Rate to expresslrs_rf_pref_params_s. These were determined by Sunjun using scatter plots and trending across millions and millions of packets of simultaneous representative RX testing across many long walks.
![image](https://user-images.githubusercontent.com/677183/163480086-741d52f9-fe38-4b52-99a9-073224a188ec.png)

### If all else fails
If no other power change has resulted from a LinkStats packet and the LQ is <85, the power is bumped one notch.

### Legacy adjustments
The boost and overload conditions were brought from the existing dynamic power system as well
* If RSSI >= -5: lower power one level to prevent overloading the RX frontend (with or without dynamic power)
* If RSSI <= -20 and not DynamicPower: return to max
* Dynamic Power on a switch: boost to max if dynamic power disabled by switch
* If Armed and Disconnected: boost to max
* If LQ drops from the moving average by 20 in a single update: boost to max
* If LQ is below 50: boost to max

## FLRC dynamic power
FLRC does not have SNR, therefore the SNR-based algorithm will not work for it. Instead it uses the old RSSI-based system, raising the power if the RSSI is below `sensitivity_limit + 15` and raising the power if RSSI is above `sensitivity_limit + 21` and LQ >= 95.

## Unrelated (?) change
* The TX LQ (the link quality reported for telemetry packets received by the TX) used to account for at most 10 packets. This has been increased to 31 since it takes the same RAM and code space.

SunjunKim commented:
Freakin darn good documentation, as always!
We've tested this for a long time, but mostly on Team2400.

More reviews&tests are required for Team900 gears

avsaase commented:
Awesome! I agree that SNR is by far the most predictive factor (that we have) of future packet LQ. 

One concern I have is how this behaves with high telemetry ratios such as 1:2 and 1:4. At these ratios, there is hardly any averaging of SNR on the receiver side. In my testing, I also found per packet SNR to be noisier at 500Hz than at lower packet rates. The result is that at for example 500Hz and a 1:4 telemetry ratio dynamic power constantly changes the power level. I think it would make sense to add some extra averaging on the TX side for these high downlink rates.

CapnBry commented:
> One concern I have is how this behaves with high telemetry ratios such as 1:2 and 1:4. At these ratios, there is hardly any averaging of SNR on the receiver side

A valid concern, but if you read the description (and the code) carefully:
> average the SNR value between each LinkStats packet

Between each LinkStats packet, not each telemetry packet. LinkStats are sent at most every 512ms regardless of TLM ratio. 

There is the exception that someone could be running 1:2 and also NOT have any other telemetry to send because they're not connected to a FC or something. This would send LinkStats more frequently because there's no other data to send. I'm not going to add more averaging code on the TX side or tracking it on the RX side just for dumb case. That user already doesn't give a shit about their link anyway, ;)

avsaase commented:
I did two test flights today, one with the EP1 on 150Hz/1:4, and the other with the Matek ELRS-R24-D on 150HZ/1:8. With the EP1 the power level was very erratic ([DVR sample](https://www.youtube.com/watch?v=QRrdrHX19Lw)). I didn't have SNR on the OSD or debug logs so I have to speculate about what is driving the power changes. 

With the Matek receiver after an LQ-based boost, the power decreases indeed seem to occur about twice a second so that seems to work correctly. That leaves the missed telemetry packets to explain the erratic power changes with the EP1. I think the proactive power increases after a LinkStats packet is missed are too aggressive with high downlink rates in combination with lower power receivers.

SunjunKim commented:
> I did two test flights today, one with the EP1 on 150Hz/1:4, and the other with the Matek ELRS-R24-D on 150HZ/1:8. With the EP1 the power level was very erratic ([DVR sample](https://www.youtube.com/watch?v=QRrdrHX19Lw)). I didn't have SNR on the OSD or debug logs so I have to speculate about what is driving the power changes.

Thanks for the testing. You may try different up/down SNR thresholds, and that checked whether that could make the problem better (actually the power-down thresholds were not tested in real flights)

CapnBry commented:
> I did two test flights today, one with the EP1 on 150Hz/1:4, and the other with the Matek ELRS-R24-D on 150HZ/1:8. With the EP1 the power level was very erratic ([DVR sample](https://www.youtube.com/watch?v=QRrdrHX19Lw)). I didn't have SNR on the OSD or debug logs so I have to speculate about what is driving the power changes.

I see what you're saying and I'll do some testing to see if there's a problem with the algorithm. The important thing to note is that if you're running 150Hz 1:8, that means that the missing telemetry bit of code has missed 9-10 telemetry packets **in a row** before the first power bump. At 1:4 that's 18-19 telemetry packets in a row that have been lost. If it is working as I've described it is doing the right thing by moving the power all over the place as the telemetry link is really terrible.

I'll take a look at it, but the dynamic power algorithm's primary purpose is to protect the link integrity, not stay at the same power level. Ironically, the lower telemetry ratios are used by people who want the fastest response out of the dynamic power system and the higher ratios are used by people who want it to move more slowly. The input is at odds with expectations! 😅

avsaase commented:
Power decreases can only occur when a LinkStats tlm packet is received, SNR is high and LQ is good. When that happens, shouldn't there be a minimum 512ms delay before power can be increased again, either because SNR was low or because the LinkStats packet was missed? After that initial delay power can be increased rapidly on every missed tlm packet. 

I know the OSD is far from the ideal way to see what's going on with the power level because it has to be sent up from the TX and it's only included every 8 (correct?) uplink packets. But still, it looks like the time between a power decrease and a subsequent power increase is shorter than it should be. Is it possible the 512ms delay times doesn't get reset when a LinkStats tlm packet is received?

CapnBry commented:
Tagging "Do not merge" for now, until I have a chance to look more closely at higher telemetry rates

JyeSmith commented:
Has anyone looked at the rx power level?  If the tx is no receiving tlm packets, the rx output might just be too low.  For receivers with adjustable output can they be commanded to go max output?

CapnBry commented:
> For receivers with adjustable output can they be commanded to go max output?

There's a receiver with adjustable power output? 🤔

pkendall64 commented:
> > For receivers with adjustable output can they be commanded to go max output?
> 
> There's a receiver with adjustable power output? 🤔

Yes boss, BETAFPV Nano (the one wit the PA).

SunjunKim commented on src/src/common.cpp:
```diff
@@ -17,10 +17,10 @@ expresslrs_mod_settings_s ExpressLRS_AirRateConfig[RATE_MAX] = {
-    {2, RATE_LORA_50HZ,  -120, 18560, 4000, 2500, 600, 5000},
-    {3, RATE_LORA_25HZ,  -123, 29950, 6000, 4000,   0, 5000}};
+    {0, RATE_LORA_200HZ, -112,  4380, 3000, 2500, 600, 5000, -1,  2},
+    {1, RATE_LORA_100HZ, -117,  8770, 3500, 2500, 600, 5000, -3,  1},
```
For 100 Hz, better to be 0, 2 here

CapnBry commented:
Everyone thinks they have the best idea about how dynamic power should work, but I am not sure there's any winning solution when SNR*4 vs FHSS hop looks like this (50,000 packets Team900 50Hz)
![image](https://user-images.githubusercontent.com/677183/172020698-39a85633-9130-4345-8f65-ce28c5c2a785.png)

Same with RSSI vs FHSS
![image](https://user-images.githubusercontent.com/677183/172020816-2a63e575-5c1a-454c-8d66-b895dc3a43bf.png)

What is the appropriate power to use when the power level is fine sometimes and not at all fine other times. In b4 "just track the signal strength per channel and avoid the ones that are poor dynamically".

CapnBry commented:
There's two things in the updated code
* The SNR value is now SNR*4 for better precision, described in OP
* I fixed a bit of an oversight with the "missing telemetry" code, if every LinkStats is missed, but all the other telemetry was received, it would keep resetting the timer and not ever increase the power. Now it only considers the time since the last LinkStats telemetry, and then up the power for each missed telemetry packet after that.
* The missed telemetry event wasn't being thrown at the end of the telemetry slot, but rather at the beginning of the following one, so I fixed that too.

avsaase commented on src/src/dynpower.cpp:
```diff
@@ -0,0 +1,204 @@
+  {
+    DBGLN("+power (lq)");
+    POWERMGNT::incPower();
+    //--powerHeadroom;
```
Shouldn't this be uncommented?

CapnBry replied:
No it is commented and left in so if any code is added after it, it will have to be returned. It is commented out just in case the complier decides it actually wants to leave that code in and costs us 4 bytes. Every byte is sacred!

pkendall64 commented on src/lib/SX12xxDriverCommon/SX12xxDriverCommon.h:
```diff
@@ -36,7 +36,8 @@ class SX12xxDriverCommon
     /////////////Packet Stats//////////
     int8_t LastPacketRSSI;
-    int8_t LastPacketSNR;
+    //int8_t LastPacketSNR;
```
Remove commented line

pkendall64 commented on src/src/dynpower.cpp:
```diff
@@ -0,0 +1,202 @@
+void ICACHE_RAM_ATTR DynamicPower_TelemetryUpdate(int8_t snrScaled)
+{
+    dynpower_updated = snrScaled;
+    // static unsigned lastUp;
+    // if (snrScaled != DYNPOWER_UPDATE_MISSED)
+    // {
+    //   unsigned now = millis();
+    //   DBGLN(" %u P=%u SNR=%d ", now - lastUp, POWERMGNT::currPower(), snrScaled);
+    //   lastUp = now;
+    // }
```
Did you want to wrap this in some DEBUG_ flag? Or just remove it?

CapnBry replied:
I'll just remove it, I didn't want to have to come up with a new DEBUG_ name and it is easy enough to just throw in DBG if needed. This already served its purpose.

pkendall64 commented on src/src/dynpower.cpp:
```diff
@@ -0,0 +1,202 @@
+      int8_t avg_rssi = dynpower_mean_rssi.mean(); // resets it too
+      if ((avg_rssi < rssi_inc_threshold) && (powerHeadroom > 0))
+      {
+        DBGLN("+power (rssi)");
```
Why is this one DBGLN and the one below DBGVLN?

pkendall64 commented on src/src/dynpower.cpp:
```diff
@@ -0,0 +1,202 @@
+    //int8_t snrScaled = CRSF::LinkStatistics.uplink_SNR;
+    if (snrScaled >= ExpressLRS_currAirRate_RFperfParams->DynpowerSnrThreshDn && lq_avg >= DYNPOWER_LQ_THRESH_DN)
+    {
+      DBGVLN("-power (snr)");
```
Also, here, DBGVLN for down, an DBGLN for up. So I assume you intend this to be the case.

pkendall64 replied:
I do remember now you mention it. When I was typing the comments I felt a little déjà vu moment in my head.
Good comments in the code, I won't ask why it's like this again 😉

pkendall64 reviewed (approved):
Tested using my shonky TX which has bad filtering and walking around the house I can see it change up and down the power levels and maintaining a good LQ.

I've added some comments, nothing should stop a merge.

pkendall64 commented:
I think would also close #1233

SunjunKim reviewed (approved):
I've been testing this since the beginning. All is functional now.

Further adjustment of SNR thresholds should be done in a separate PR later.

fireduckfpv commented:
Great work! :) 
I was just wondering, would it help to have dynamic Hz as well? That way one could just set the Hz to 200, and it would bump it down when needed?

CapnBry commented:
> I was just wondering, would it help to have dynamic Hz as well?

No, we do not have a reliable way to change air rates on the fly and trying to do it under dicey conditions is even more dangerous and not going to happen on a single link system.
